clang-tools-extra/clangd/InlayHints.cpp

```diff
@@ -372,6 +374,292 @@ maybeDropCxxExplicitObjectParameters(ArrayRef<const ParmVarDecl *> Params) {
+      addLabel(Prefix.str());
+  }
+
+  void VisitType(const Type *) { addLabel(CurrentType.getAsString(PP)); }
+
+  void VisitTagType(const TagType *TT) {
+    auto *D = TT->getDecl();
+    if (auto *Specialization = dyn_cast<ClassTemplateSpecializationDecl>(D))
+      return handleTemplateSpecialization(
+          TemplateName(Specialization->getSpecializedTemplate()),
+          Specialization->getTemplateArgs().asArray());
+    if (auto *RD = dyn_cast<CXXRecordDecl>(D);
+        RD && !RD->getTemplateInstantiationPattern())
+      return addLabel(
+          [&](llvm::raw_ostream &OS) { return RD->printName(OS, PP); },
+          [&] { return nameLocation(*RD, SM); });
+    return VisitType(TT);
+  }
+
+  void VisitEnumType(const EnumType *ET) {
+    return addLabel(
+        [&](llvm::raw_ostream &OS) { return ET->getDecl()->printName(OS, PP); },
+        [&] { return nameLocation(*ET->getDecl(), SM); });
+  }
+
+  void VisitAutoType(const AutoType *AT) {
+    if (!AT->isDeduced() || AT->getDeducedType()->isDecltypeType())
+      return;
+    maybeAddQualifiers();
+    CurrentTypeRAII Guard(*this, AT->getDeducedType());
+    return Visit(AT->getDeducedType().getTypePtr());
+  }
+
+  void VisitElaboratedType(const ElaboratedType *ET) {
+    maybeAddQualifiers();
+    if (auto *NNS = ET->getQualifier()) {
+      switch (NNS->getKind()) {
+      case NestedNameSpecifier::Identifier:
+      case NestedNameSpecifier::Namespace:
+      case NestedNameSpecifier::NamespaceAlias:
+      case NestedNameSpecifier::Global:
+      case NestedNameSpecifier::Super: {
+        std::string Label;
+        llvm::raw_string_ostream OS(Label);
+        NNS->print(OS, PP);
+        addLabel(std::move(Label));
+      } break;
+      case NestedNameSpecifier::TypeSpec:
+      case NestedNameSpecifier::TypeSpecWithTemplate:
+        CurrentTypeRAII Guard(
+            *this,
+            QualType(
+                NNS->getAsType(),
+                /*Quals=*/0) // Do we need cv-qualifiers on type specifiers?
+        );
+        Visit(NNS->getAsType());
+        addLabel("::");
+        break;
+      }
+    }
+    CurrentTypeRAII Guard(*this, ET->getNamedType(), ET->getQualifier());
+    return Visit(ET->getNamedType().getTypePtr());
+  }
+
+  void VisitReferenceType(const ReferenceType *RT) {
+    maybeAddQualifiers();
+    QualType Next = skipTopLevelReferences(RT->getPointeeTypeAsWritten());
+    CurrentTypeRAII Guard(*this, Next);
+    Visit(Next.getTypePtr());
+    if (Next->getPointeeType().isNull())
+      addLabel(" ");
+    if (RT->isLValueReferenceType())
+      addLabel("&");
+    if (RT->isRValueReferenceType())
+      addLabel("&&");
+  }
+
+  void VisitPointerType(const PointerType *PT) {
+    QualType Next = PT->getPointeeType();
+    std::optional<CurrentTypeRAII> Guard(std::in_place, *this, Next);
+    Visit(Next.getTypePtr());
+    if (Next->getPointeeType().isNull())
+      addLabel(" ");
+    addLabel("*");
+    Guard.reset();
```

Comment:
Instead of using `optional`, I think this would be clearer:

```c++
    QualType Next = PT->getPointeeType();
    {
      CurrentTypeRAII Guard(*this, Next);
      Visit(Next.getTypePtr());
      if (Next->getPointeeType().isNull())
        addLabel(" ");
      addLabel("*");
    }
    maybeAddQualifiers();
```

---

And with this change, I now notice that all uses of `CurrentTypeRAII` follow this pattern:

```c++
// Q is a QualType
CurrentTypeRAII Guard(*this, Q, /* maybe an NNS */);
Visit(Q.getTypePtr());
```

What do you think about replacing `CurrentTypeRAII` with a function `VisitQualType` like this:

```c++
void VisitQualType(QualType Q, NestedNameSpecifier *NNS = nullptr) {
  QualType PreviousType = CurrentType;
  NestedNameSpecifier* PreviousNNS = CurrentNestedNameSpecifier;
  CurrentType = Q;
  CurrentNestedNameSpecifier = NNSl

  TypeVisitor::Visit(Q.getTypePtr());

  CurrentType = PreviousType;
  CurrentNestedNameSpecifier = PreviousNNS;
}
```

As a bonus, we can also remove the `Current` constructor parameter, and make the usage of the builder like this:

```c++
TypeHintBuilder Builder(/* other arguments */);
Builder.VisitQualType(Q);  // what we would have passed as Current to the constructor
```

These changes can help avoid mistakes where by accident the type stored in `CurrentType` and the type passed to `TypeVisitor::Visit()` are different.

Reply:
Great! This indeed simplifies the code a lot.
(I used an RAII object because that's a typical pattern in Sema/Parser. :)